open_spiel/games/dou_dizhu/dou_dizhu_utils.cc

```diff
@@ -0,0 +1,860 @@
+  else if(rank == kNumRanks - 2) return "(BWJ)";
+  else if(rank == kNumRanks - 1) return "(CJ)";
+  else SpielFatalError("Non valid rank");
+}
+
+std::string FormatSingleHand(const std::array<int, kNumRanks>& hand){
+  std::string hand_format;
+  for (int rank = 0; rank < kNumRanks; ++rank){
+    for (int i = 0; i < hand[rank]; ++i) absl::StrAppend(&hand_format, RankString(rank));
+  }
+  return hand_format;
+}
+
+
+
+// resolve ambiguity for cases like 333444555666
+std::string FormatAirplaneCombHand(int action){
+  TrioCombParams params = GetAirplaneCombParams(action);
+  std::array<int, kNumRanks> hand = ActionToHand(action);
+  std::string airplane_comb_str;
+  // specify which is chain
+  for(int rank = params.chain_head; rank < params.chain_head + params.chain_length; ++rank){
+    for(int i = 0; i < 3; ++i) absl::StrAppend(&airplane_comb_str, RankString(rank));
+  }
+  absl::StrAppend(&airplane_comb_str, "-");
+  // kickers
+  for(int rank = 0; rank < kNumRanks; ++rank){
+    if(rank >= params.chain_head && rank < params.chain_head + params.chain_length) continue;
+    if(!hand[rank]) continue;
+    for(int i = 0; i < hand[rank]; ++i) absl::StrAppend(&airplane_comb_str, RankString(rank));
+  }
+  return airplane_comb_str;
+}
+
+
+// Shared by single-rank and chain-only hands
+int GetNumCardsPerRank(int action){
+  int num_cards;
+  if(action >= kPlayActionBase && action < kPairActionBase) {
+    num_cards = 1;
+  } else if(action >= kPairActionBase && action < kTrioActionBase) {
+    num_cards = 2;
+  } else if((action >= kTrioActionBase && action < kTrioWithSoloActionBase) || 
+        (action >= kAirplaneActionBase && action < kAirplaneWithSoloActionBase)) {
+    num_cards = 3;
+  } else if(action >= kBombActionBase && action < kRocketActionBase) {
+    num_cards = 4;
+  } else SpielFatalError("Invalid action ID");
+
+  return num_cards;
+}
+
+
+int GetSingleRankActionBase(int num_cards_same_rank=1){
+  int action_base;
+  switch (num_cards_same_rank){
+  case 1:
+    action_base = kPlayActionBase;
+    break;
+  case 2:
+    action_base = kPairActionBase;
+    break;
+  case 3:
+    action_base = kTrioActionBase;
+    break;
+  case 4:
+    action_base = kBombActionBase;
+    break;
+  default:
+    SpielFatalError("The number of cards of the same rank is wrong (single rank).");
+  }
+  return action_base;
+}
+
+
+
+SingleRankHandParams GetSingleRankHandParams(int action){
+  const int num_cards = GetNumCardsPerRank(action);
+  const int action_base = GetSingleRankActionBase(num_cards);
+  SPIEL_CHECK_GE(action, action_base);
+  return SingleRankHandParams(action - action_base, num_cards);
+}
+
+
+std::array<int, kNumRanks> SingleRankHand(int action){
+  std::array<int, kNumRanks> hand{};
+  SingleRankHandParams params = GetSingleRankHandParams(action);
+  hand[params.rank] = params.num_cards;
+  return hand;
+}
+
+// given a single-rank hand, map it to action id
+int SingleRankHandToActionId(const std::array<int, kNumRanks>& hand){
```

Comment:
Another example of Span<const int>, as you're not modifying the hand container (or any of the members).